daemon/api_sideload_n_try.go

```diff
@@ -207,6 +212,118 @@ out:
 	return AsyncResponse(nil, chg.ID())
 }
 
+// maxReadBuflen is the maximum buffer size for reading the non-file parts in the snap upload form
+const maxReadBuflen = 1024 * 1024
+
+// readForm returns a Form populated with values (for non-file parts) and file headers (for file
+// parts). The file headers contain the original file name and a path to the persisted file in
+// dirs.SnapDirBlob. If an error occurs and a non-nil Response is returned, an attempt is made
+// to remove temp files.
+func readForm(reader *multipart.Reader) (_ *Form, apiErr *apiError) {
+	availMemory := int64(maxReadBuflen)
+	form := &Form{
+		Values:   make(map[string][]string),
+		FileRefs: make(map[string][]*FileReference),
+	}
+
+	// clean up if we're failing the request
+	defer func() {
+		if apiErr != nil {
+			form.RemoveAll()
+		}
+	}()
+
+	for {
+		part, err := reader.NextPart()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return nil, BadRequest("cannot read POST form: %v", err)
+		}
+
+		name := part.FormName()
+		if name == "" {
+			continue
+		}
+
+		filename := part.FileName()
+		if filename == "" {
+			// non-file parts are kept in memory
+			buf := &bytes.Buffer{}
+
+			// copy one byte more than the max so we know if it exceeds the limit
+			n, err := io.CopyN(buf, part, availMemory+1)
+			if err != nil && !errors.Is(err, io.EOF) {
+				return nil, BadRequest("cannot read form data: %v", err)
+			}
+
+			availMemory -= n
+			if availMemory < 0 {
+				return nil, BadRequest("cannot read form data: exceeds memory limit")
+			}
+
+			form.Values[name] = append(form.Values[name], buf.String())
+			continue
+		}
+
+		tmpPath, err := writeToTempFile(part)
+
+		// add it to the form even if err != nil, so it gets deleted
+		ref := &FileReference{TmpPath: tmpPath, Filename: filename}
+		form.FileRefs[name] = append(form.FileRefs[name], ref)
```

Comment:
So I wonder if it's possible to procure a request which sends out multiple files using different part names, and then another one which sends multiple parts named "snap". In which case, I think we would save all those files to temporaries, but then we only consume one and never clean up the rest. Can you add test for this?

Reply:
Very true. I fixed it now. Thanks